Source/WebCore/dom/WindowEventLoop.cpp

```diff
@@ -158,21 +158,52 @@ bool WindowEventLoop::shouldEndIdlePeriod()
 
 MonotonicTime WindowEventLoop::computeIdleDeadline()
 {
+    auto minTime = m_lastIdlePeriodStartTime + 50_ms;
+
+    auto timerTime = nextTimerFiringTime();
+    if (timerTime && *timerTime < minTime)
+        minTime = *timerTime;
+
     if (!m_pagesWithRenderingOpportunity.isEmptyIgnoringNullReferences()) {
-        MonotonicTime minRenderingUpdateTime = MonotonicTime::infinity();
         for (auto it : m_pagesWithRenderingOpportunity) {
-            if (it.value < minRenderingUpdateTime)
-                minRenderingUpdateTime = it.value;
+            if (it.value < minTime)
+                minTime = it.value;
         }
-        return minRenderingUpdateTime;
     }
-    return m_lastIdlePeriodStartTime + 50_ms;
+
+    return minTime;
 }
 
 void WindowEventLoop::didReachTimeToRun()
 {
     Ref protectedThis { *this }; // Executing tasks may remove the last reference to this WindowEventLoop.
     run();
+
+    if (hasTasksForFullyActiveDocument() || !microtaskQueue().isEmpty())
+        return;
+
+    bool hasIdleCallbacks = false;
+    forEachAssociatedContext([&](ScriptExecutionContext& context) {
+        RefPtr document = dynamicDowncast<Document>(context);
+        if (!document)
+            return;
+        auto* idleCallbackController = document->idleCallbackController();
+        if (!idleCallbackController)
+            return;
+        hasIdleCallbacks = hasIdleCallbacks || !idleCallbackController->isEmpty();
+    });
```

Comment:
Nit - I suppose something like a `findAssociatedContextMatching(…)` might be _ever so slightly_ more efficient here, since we can stop iteration early once we find non-empty idle callbacks (though the list of associated contexts is likely pretty short anyways).

Reply:
Sure, added findMatchingAssociatedContext to EventLoop.cpp.